### ISSUE TITLE
False positive ArgumentTypeCoercion with unsealed array shapes

### ISSUE THREAD

**Issue by mstilkerich** [vimeo/psalm#8980]
Hello,

  when I (read) access an optional key in an unsealed array shape, psalm loses the type information on the accessed key value. The same code is accepted without problems by psalm 4.

The problem does not occur if the array shape is sealed. It also doesn't occur if the accessed key is not optional.

https://psalm.dev/r/a420ace90a

```php
/**
 * @psalm-type SaveData = array{name?: string, ...}
 */
class Test {
    /**
     * @param SaveData $sd
     * @return SaveData
     */
    public function foo(array $sd): array
    {
        return $sd;
    }

    /**
     * @param SaveData $sd
     */
    public function bar(array $sd) : void
    {
        // This line causes the psalm error, psalm assumes type array{name?: mixed|string, ...<array-key, mixed>} for $sd after this line
        echo ($sd["name"] ?? "no name");
        $this->foo($sd);
    }    
}
```

**Comment by mstilkerich** [comment]
Probably the same issue:
https://psalm.dev/r/d6e309af74

The isset() appears to cause a change of the type of the optional array element.

**Comment by christianseel** [comment]
I'm running into the same issue with Psalm v6.8.9.

It's not only related to an `isset` call, any usage of the array seems to cause the issue: https://psalm.dev/r/ea0ecc8263

The type `mixed` for the unsealed keys is merged with the type of the optional key, which is unexpected behavior for me.
